### PR TITLE
New layout blocks for sprite lab

### PR DESCRIPTION
**Pull request by bencodeorg**
Adds new sprite lab blocks with same behavior to existing blocks in Dance Party (albeit with a slightly shorter list of possible layouts). All logic in this PR was derived from our dance party repo, with some light massaging/organizing to get things to work in sprite lab.

| Existing Dance Party Block      | New Sprite Lab Block |
| ----------- | ----------- |
| image      | image       |
| image   | image        |

Screen Recording 2023-04-05 at 5 51 00 PM

## Links

- jira ticket: [SL-694](https://codedotorg.atlassian.net/browse/SL-694)
- [logic from dance party repo](https://github.com/code-dot-org/dance-party/blob/6672bdb0cffad1cbfda6e7396155f542b6cdcffe/src/p5.dance.js#L637-L911)
- [tests from dance party repo](https://github.com/code-dot-org/dance-party/blob/main/test/unit/layoutTest.js)

## Testing story

Tested manually (see video above). Copied over relevant unit tests from dance party repo.

### CODE REVIEW

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js:18]: The sprites look bigger than expected after the layout is applied. Can we this step to resize each sprite? https://github.com/code-dot-org/dance-party/blob/6672bdb0cffad1cbfda6e7396155f542b6cdcffe/src/p5.dance.js#L641

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js:30]: Nit: Would using a `switch` to check the layout type improve readability?  Some of the specific layout types have their own if/else-if/else logic already.

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js:145]: Could this be just be `group.forEach (sprite => {})`?

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js:160]: Another place that could use `group.forEach`?

File: apps/src/p5lab/spritelab/commands/spriteCommands.js
Comment (mikeharv) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js:183]: Is there an advantage to using `.concat` over `.push` here?

File: dashboard/config/blocks/GamelabJr/gamelab_makeNewSpriteGroup.json
Comment (mikeharv) [inline dashboard/config/blocks/GamelabJr/gamelab_makeNewSpriteGroup.json:6]: Thanks for adding styles to the blocks! This will be especially helpful once we migrate Sprite Lab to Google Blockly. In the meantime, because CDO Blockly doesn't support styles like this, we'll want to also add a hard-coded HSV value so that the block gets the expected red color.

You should just need to add this:
```
  "color": [
    355,
    ".7",
    ".7"
  ],
```

(I don't have have these values memorized, so I just checked another sprite block from the same pool.)

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js:107]: Just noting that it looks like the "circle" format has its own logic for scale that overrides the `30` constant set at the top of the dance version of `layoutSprite`.

File: apps/test/unit/p5lab/spritelab/actionCommandsTest.js
Comment (mikeharv) [inline apps/test/unit/p5lab/spritelab/actionCommandsTest.js:543]: It looks like we do test that the sprite re-sizing works as expected. Should we consider adding a simple check to one of the other layouts to make sure that the sprites get resized for it as well? Again, following the Dance logic, I think we should set each sprite's scale to `30` for all layouts besides "circle".

**Review by mikeharv — COMMENTED** [review]
Great to see this working so smoothly! I really like the result of putting most of the logic in `layoutUtils.js` and keeping the actual command shortcuts nice and short.

It would be great to get an updated gif that shows the red block color for the new `make` block as well as any resizing that we add.

The tests should cover this, but I also couldn't tell if the "circle" layout was dynamically resizing the group based on the number of sprites in the group or not.

Do you think it would make sense to add follow-up task to the backlog to move the other dance layouts into `layoutUtils.js`? This could make it simpler to enable them as options for Sprite Lab, if that was ever desired, and it could allow us to remove some of the original code from the Dance repo. It's probably not urgent work, but I feel like it could be a nice maintenance task, perhaps for someone looking to become familiarized with our various labs and separate repos.

File: apps/src/p5lab/layoutUtils.js
Comment (bencodeorg) [reply apps/src/p5lab/layoutUtils.js:18]: ooh good catch, will update!

File: apps/test/unit/p5lab/spritelab/actionCommandsTest.js
Comment (bencodeorg) [reply apps/test/unit/p5lab/spritelab/actionCommandsTest.js:543]: Yes! Would have caught the sizing oversight, will add

File: apps/src/p5lab/layoutUtils.js
Comment (bencodeorg) [reply apps/src/p5lab/layoutUtils.js:30]: I decided against ES6-ifying/cleaning up all of this for a couple reasons, LMK if you strongly disagree:
- if future improvements are needed across both dance and spritelab, will require developers to have to translate between ES6 and older syntax -- leaving as-is makes making changes in the same places easier to identify/make. If we want to refactor, we should refactor both
- feels out of scope for this work to just add this functionality in spritelab to also refactor it

**Comment by bencodeorg** [comment]
> Do you think it would make sense to add follow-up task to the backlog to move the other dance layouts into `layoutUtils.js`? This could make it simpler to enable them as options for Sprite Lab, if that was ever desired, and it could allow us to remove some of the original code from the Dance repo. It's probably not urgent work, but I feel like it could be a nice maintenance task, perhaps for someone looking to become familiarized with our various labs and separate repos.

I think that would be great, and would be a nice opportunity to do some of the ES6 stuff you suggested. I don't understand the split between dance party and sprite lab well enough to know whether this is a relatively easy refactor, or something that would be straightforward. I can add something to the backlog though to investigate at least?

File: dashboard/config/blocks/GamelabJr/gamelab_makeNewSpriteGroup.json
Comment (bencodeorg) [reply dashboard/config/blocks/GamelabJr/gamelab_makeNewSpriteGroup.json:6]: The layout block should just stay blue, yeah? Red = creates sprites?

**Comment by bencodeorg** [comment]
> The tests should cover this, but I also couldn't tell if the "circle" layout was dynamically resizing the group based on the number of sprites in the group or not.

There is dynamic resizing for the circle layout, but its a little funky because it seems like the sprites are different sizes (eg, ladybug/bee vs cow/elephant) so it looks a lot different based on the sprite. Maybe this is kind of desired (small vs large animals have sprites of corresponding size)?

circle layout with sprite and count comparison

File: apps/src/p5lab/layoutUtils.js
Comment (jmkulwik) [inline apps/src/p5lab/layoutUtils.js:10]: Would it make sense to reuse some of that logic rather than copy it over here? Perhaps making a common layoutUtils file?

Comment (jmkulwik) [reply apps/src/p5lab/layoutUtils.js:10]: Then it may be easier to apply code updates such as those suggested below.

File: apps/src/p5lab/layoutUtils.js
Comment (jmkulwik) [reply apps/src/p5lab/layoutUtils.js:30]: I think reusing a the dance code and making a common layouts file could help a lot here. One thing about how this is written that makes me nervous is the large chunks of code between the main if/else statements. Whether it moves to a switch or not, it would be good to move the logic into separate functions so the layoutSpriteGroup function is only handling which function to call. (i.e. I'm thinking about the case where a future maintainer misplaces a curly brace and ends up functionally removing a bunch of layouts accidentally)

File: apps/src/p5lab/spritelab/commands/spriteCommands.js
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/commands/spriteCommands.js]: What's n in this case? It may clear up some confusion to add a bit more context in the variable name. For example `spriteCount` or `numGroups` etc.

**Comment by mikeharv** [comment]
> > The tests should cover this, but I also couldn't tell if the "circle" layout was dynamically resizing the group based on the number of sprites in the group or not.
> 
> There is dynamic resizing for the circle layout, but its a little funky because it seems like the sprites are different sizes (eg, ladybug/bee vs cow/elephant) so it looks a lot different based on the sprite. Maybe this is kind of desired (small vs large animals have sprites of corresponding size)?
> 
> circle layout with sprite and count comparison [ circle layout with sprite and count comparison ](https://user-images.githubusercontent.com/25372625/229642929-1c30846f-2a25-4bfb-a490-227a7c79d191.gif) [ ](https://user-images.githubusercontent.com/25372625/229642929-1c30846f-2a25-4bfb-a490-227a7c79d191.gif)

I forgot that this might happen! It's unfortunately not what we want, but it's an easy fix. In Sprite Lab, we apply some special logic to autoscale sprites based on the actual image dimensions of the animation/costume: https://github.com/code-dot-org/code-dot-org/blob/ben/sprite-lab-layout-block/apps/src/p5lab/spritelab/CoreLibrary.js#L438-L457

This mean that instead of setting `sprite.scale` directly, we want to use `sprite.setScale()`. There is also a `sprite.getScale()` for reading the value. I think we just need to update this in 3 places, which I'll flag individually in case it's helpful. (I tested this locally so I could be sure I was giving the right advice.)

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js]: Use `sprite.setScale(0.3)` here.

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js]: sprite.scale(scale)

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js]: replace `sprite.scale` with `sprite.getScale()`

File: apps/src/p5lab/spritelab/commands/spriteCommands.js
Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: We usually use `numSprites` for this argument in similar commands (see `makeNumSprites` or `makeBurst` defined above at lines 68 and 83).

File: apps/test/unit/p5lab/spritelab/actionCommandsTest.js
Comment (mikeharv) [reply apps/test/unit/p5lab/spritelab/actionCommandsTest.js:543]: Note that we'll probably want to move to `getScale()` here as well as throughout this file. It's possible the tests will still pass because the test sprites aren't using actual animations with different dimensions.

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [reply apps/src/p5lab/layoutUtils.js:10]: The old logic is on the `dance-party` repo, so it does feel like the refactor we all want would greatly increase the scope of this task. Given that the curriculum team needs these blocks in order to start building new lessons, I'm hesitant to do that right away. I do think it's a valuable next step. @bencodeorg has already laid some good groundwork for this by creating this file. Next steps would be to:
* copy all other layout logic from api in the `dance-party` repo into the main repo
* add lab-specific logic for how sprites get resized (see this comment: https://github.com/code-dot-org/code-dot-org/pull/51061#issuecomment-1497490196)
* somehow get dance to use the new utils instead of what's in the old api
* remove the old logic from the api

Meta-question about this: Much of Dance Party's inner workings seem to be self-contained within its own repo. I don't know if anyone besides us actually uses it as a plugin, but would it be bad to remove this logic from the API and instead make it rely on the cdo repo?

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js]: After playing with this locally, I think actually be unnecessary to include it for the Sprite Lab version of layouts. In Dance Party, we magically play with the depth of sprites to give a sort of  3D foreground vs. background (smaller sprites appear to be further back). In Sprite Lab, depth logic is really only handled by the order in which the sprites are created.

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [inline apps/src/p5lab/layoutUtils.js]: After playing with this locally, I don't believe we want to adjust the rotation of the sprites if we're in Sprite Lab. Dance Party's sprites are all anthropomorphic/humanoid. Students don't actually program the sprites to move around in the same way they do in Sprite Lab. Messing with the rotation could create some false impressions of how movement will be impacted.

Here's a comparison of how it works right now, with how it would work if we kept the rotation at 0. Note that after creating the two groups, each is told to move a different direction using a behavior.

rotated circle sprites
not rotated

Comment (mikeharv) [reply apps/src/p5lab/layoutUtils.js]: Going a step further, I think that for Sprite Lab, we don't want to even set the rotation to 0 and instead just leave it what it was originally. This won't affect sprites that were just created, but if the blue layout block is used in conjunction with events, re-initializing the sprite rotation leads to undesired  out comes if the sprites are moving around and changing directions.

In this example, note that the rocket ships seem to be able to be turned and driven forward using the arrow keys, but only until the layout is changed into a circle. Once that happens, the movement direction is messed up.


https://user-images.githubusercontent.com/43474485/230114097-ca4faafb-1171-4dc1-842f-9be98f136571.mp4

**Review by mikeharv — COMMENTED** [review]
I pulled the changes and spend some time testing them. I think my suggested tweaks are all easy single line changes. Basically though, programming Sprite Lab is different enough from Dance Party that the layouts need to work slightly differently when it comes to each sprite's scale, depth and rotation:

* Scale - use `setScale` and `getScale` instead of setting the property directly
* Depth - don't adjust the depth of sprites
* Rotation - don't adjust the rotation of sprites

I also strongly agree with your plan to hold on general refactoring for now so that we can deliver these blocks (in a good working state) to the curriculum team this sprint.

File: apps/src/p5lab/layoutUtils.js
Comment (bencodeorg) [reply apps/src/p5lab/layoutUtils.js:10]: +1 to Mike's comments -- I think a refactor is a great goal, but seems out of scope given time pressure on curriculum side?

File: apps/src/p5lab/spritelab/commands/spriteCommands.js
Comment (bencodeorg) [reply apps/src/p5lab/spritelab/commands/spriteCommands.js]: 👍 can do!

File: apps/src/p5lab/layoutUtils.js
Comment (bencodeorg) [reply apps/src/p5lab/layoutUtils.js:30]: Moved core logic to helper functions, think that's an improvement even if we don't do full dance party refactor right now. LMK what you think!

**Comment by bencodeorg** [comment]
@mikeharv thanks for the detailed review, learned a lot about spritelab today! I removed the depth and rotation-altering code, and used setScale as you suggested. See the updated GIF in the description for current behavior.

@jmkulwik I broke up some of the logic, but didn't fully refactor across spritelab/dance party for the reasons Mike outlined. Here's a Jira task to follow up, LMK if there's more you think we should do before merging this!

https://codedotorg.atlassian.net/browse/SL-770

**Comment by bencodeorg** [comment]
One last question -- do we actually want the scaling any time we layout sprites (it seems like it'd be more appropriate to tie scaling to the # of sprites on the screen)? It feels a little weird now since a similar block without a layout will produce sprites of a different size:

image

**Comment by bencodeorg** [comment]
btw, I "hid" some of your comments just so I could keep track of whether I had responded to them in some form. Feel free to unhide if further discussion needed!

**Comment by mikeharv** [comment]
> One last question -- do we actually want the scaling any time we layout sprites (it seems like it'd be more appropriate to tie scaling to the # of sprites on the screen)? It feels a little weird now since a similar block without a layout will produce sprites of a different size:
> 
> image

First off, are the teal blocks in your screenshot related to your question in any way? They both have the brown bunny costume selected, so they won't actually do anything here. Assuming they were just there incidentally.

A couple thoughts: 
The existing `makeNumSprites` block creates n sprites of size 100 at random locations. We shouldn't change how that block works, but should note that the new blocks were requested based on its limitations. I would not expect curriculum levels to highlight this older block and the new blocks together, but we do need to make sure they "play nicely" together according to some rules. For the sake of consistency across labs, I think we want to maintain the following:
* The new `makeNewSpriteGroup` block should not control the properties of any sprites except for those it just created, nor does it need to have awareness of other previously created sprites.
* When the `layoutSprites` **block** is used, it should control the properties of _all_ sprites with the selected costume.

These are both consistent with how Dance Party works:
image
image

In other words, while your app's preview frame maybe looks undesirable, I think it's also totally expected for that particular program. It's clear to me which sprites were created from which block. If a student wanted to further control the properties of all the sprites, ie. unify them, they can do that with any number of teal blocks, such as `layoutSprites` or `setProp`.

File: apps/src/p5lab/layoutUtils.js
Comment (mikeharv) [reply apps/src/p5lab/layoutUtils.js:10]: Additionally, as we've refined these functions for Sprite Lab (e.g. using a helper to set scale, not updating depth or rotation), we've increased the difference between versions. This doesn't make it impossible to have both labs share utilities, but it does mean we'd have to be more careful. And I definitely wouldn't want to proceed until we had time to discuss as a team which repository the logic belongs in and why.

**Review by mikeharv — APPROVED** [review]
Thanks for being willing to make so many little adjustments! It's a good reminder for us that even if it seems like you can simply borrow/copy some logic from across our P5 labs, there's still enough differences between them that we should expect it to be a bit more involved of a task. I just reviewed all of the changes since my last visit and everything looks exactly how I would expect. I think it would be great to try to merge this today (or ASAP) so that the blocks will be available on levelbuilder by Monday.

**Comment by bencodeorg** [comment]
> First off, are the teal blocks in your screenshot related to your question in any way? They both have the brown bunny costume selected, so they won't actually do anything here. Assuming they were just there incidentally. 

Sorry, bad screenshot. Brown bunny costume not relevant to my question, was just wondering about the difference between the existing create sprites block and the new create sprites in layout block, which you answered perfectly. Thanks!

**Pull request merged**
